### PR TITLE
loops-simple_programs_with_repetition-hard-q6

### PR DESCRIPTION
**Pull request by SamamthaH**
## PR Checklist
- [ ]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [ ]  Are the assertion fail messages good?
    - [ ]  Logical (does it make sense?)
    - [ ]  Readable (can a student understand what’s wrong?)
    - [ ]  Not too feed-y (does it give away the answer?)
    - [ ]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [ ]  Specific (does the fail message relate to the question?)
- [ ]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [ ]  EOF newlines
- [ ]  Do the comments for Parsons follow the correct convention?
- [ ]  Is the question being tested sufficiently?
    - [ ]  Are there representative cases to test the program?
    - [ ]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [ ]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”
---
## Website
- [ ] Is the question text on the Gamification document up to date? 
    > Make sure to include the sample output if there is one for the question

### CODE REVIEW

File: src/loops/simple_programs_with_repitition/hard/q6/MainTest.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+                new NewLine(),
+                new StringLiteral("The new query string is: "),
+                new PlaceHolder()
+        };
+    }
+
+    public void runMain(){QueryStr.main(new String[0]);}
+
+    static Stream<Arguments> inputProvider(){
+        return Stream.of(Arguments.of(" hello darkness  my old friend ", "%20%hello%20%darkness%20%%20%my%20%old%20%friend%20%"), Arguments.of("I'mAtAPayphone", "I'mAtAPayphone"), Arguments.of("   ", "%20%%20%%20%"));
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repitition/hard/q6/MainTest.java:32]: Can you try and test using an empty String? System.lineSeperator() might work

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/loops/simple_programs_with_repitition/hard/q6/MainTest.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+public class MainTest extends BaseTest {
+    // Java
+    public Clause[] testSentence(){
+        TestOption.isInputTest = true;
+        TestOption.defaultInput = " Hello Darkness My old friend ";
+        return new Clause[]{
+                new StringLiteral("Enter a sentence:"),
+                new Optional(new StringLiteral(" ")),
+                new NewLine(),
+                new StringLiteral("The new query string is: "),
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repitition/hard/q6/MainTest.java:26]: I would add the optional here as well

File: src/loops/simple_programs_with_repitition/hard/q6/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    public void runMain(){QueryStr.main(new String[0]);}
+
+    static Stream<Arguments> inputProvider(){
+        return Stream.of(Arguments.of(" hello darkness  my old friend ", "%20%hello%20%darkness%20%%20%my%20%old%20%friend%20%"), Arguments.of("I'mAtAPayphone", "I'mAtAPayphone"), Arguments.of("   ", "%20%%20%%20%"), Arguments.of(System.lineSeparator(), ""));
+    }
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    void correctlyReplacesSpaces(String in, String replaced) throws InvalidClauseException {
+        TestOption.incorrectStructureErrorMessage = "Your program does not correctly replace all spaces with %20%.";
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repitition/hard/q6/MainTest.java:40]: I would change the error message to something a bit more generic, that way you can customize what the replacement is on the website.

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/loops/simple_programs_with_repitition/hard/q6/MainTest.java (hunk 2 above)
Comment (SamamthaH) [reply src/loops/simple_programs_with_repitition/hard/q6/MainTest.java:26]: I don't know, the sample has the space obviously there, I'm not sure the optional is necessary

File: src/loops/simple_programs_with_repitition/hard/q6/MainTest.java (hunk 3 above)
Comment (SamamthaH) [reply src/loops/simple_programs_with_repitition/hard/q6/MainTest.java:40]: it'll be easy for me to just replace the %20% with a choice on the website named the same thing so it doesn't really change the amount of work I have to do.

Comment (Hedgemon4) [reply src/loops/simple_programs_with_repitition/hard/q6/MainTest.java:40]: I guess I think this pretty much tells them exactly the answer and should be more generic. 
```suggestion
        TestOption.incorrectStructureErrorMessage = "Your program does not correctly generate the query string.";
```

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/loops/simple_programs_with_repitition/hard/q6/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+    public void runMain(){QueryStr.main(new String[0]);}
+
+    static Stream<Arguments> inputProvider(){
+        return Stream.of(Arguments.of(" hello darkness  my old friend ", "%20%hello%20%darkness%20%%20%my%20%old%20%friend%20%"), Arguments.of("I'mAtAPayphone", "I'mAtAPayphone"), Arguments.of("   ", "%20%%20%%20%"), Arguments.of(System.lineSeparator(), ""));
+    }
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    void correctlyReplacesSpaces(String in, String replaced) throws InvalidClauseException {
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repitition/hard/q6/MainTest.java:39]:
```suggestion
    void generatesQueryStringCorrectly(String in, String replaced) throws InvalidClauseException {
```

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — APPROVED** [review]

**Pull request merged**